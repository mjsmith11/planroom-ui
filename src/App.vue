<template>
  <div id="app">
  <nav class="navbar navbar-expand-lg navbar-dark">
    <router-link :to="brandLink" class="navbar-brand" tag="a">
      Planroom
    </router-link>
    <button v-if="$store.getters.isLoggedIn" class="navbar-toggler" type="button" @click = "showNav = !showNav" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div v-if="$store.getters.isLoggedIn" :class="{show: showNav}" class="collapse navbar-collapse"  id="navbarSupportedContent">
      <ul class="navbar-nav mr-auto">
        <li class="nav-item" :class = "{ active: onListRoute }">
          <router-link to='/jobs' class="nav-link" tag="a">
            Job List
          </router-link>
        </li>
        <li class="nav-item" :class = "{ active: onNewRoute }">
          <router-link to='/jobs/new' class="nav-link" tag="a">
            New Job
          </router-link>
        </li>
      </ul>
      <ul class="navbar-nav">
        <li class="nav-item">
          <div class="nav-link" id = 'logout-link' @click="logout">Logout</div>
        </li>
      </ul>
    </div>
  </nav>
    <router-view/>
  </div>
</template>
<script>
import axios from 'axios'
export default {
  data () {
    return {
      showNav: false
    }
  },
  computed: {
    brandLink () {
      if (this.$store.getters.isLoggedIn) {
        return '/jobs'
      } else {
        return '/login'
      }
    },
    onListRoute () {
      return this.$route.name === 'jobs index'
    },
    onNewRoute () {
      return this.$route.name === 'new job'
    }
  },
  created () {
    axios.interceptors.response.use(undefined, function (err) {
      return new Promise(function (resolve, reject) {
        if (err.status === 401 && err.config && !err.config.__isRetryRequest) {
          this.$store.dispatch('logout')
        }
        throw err
      })
    })
  },
  methods: {
    logout () {
      this.$store.dispatch('logout')
      this.$router.push('/login')
    }
  }
}
</script>
<style>
nav{
  margin-bottom: 25px;
  background-color: #101063;
}
img {
  height: 50px;
}
div.nav-link {
  cursor: pointer;
}
</style>
