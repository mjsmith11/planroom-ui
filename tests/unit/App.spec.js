import { mount, createLocalVue } from '@vue/test-utils'
import App from '@/App'
import Vuex from 'vuex'
import axios from 'axios'

const localVue = createLocalVue()

localVue.use(Vuex)

describe('App', () => {
  it('has correct html structure logged in', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => true
    }
    let store = new Vuex.Store({
      getters
    })
    const $route = {
      name: 'test'
    }
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue,
      mocks: {
        $route
      }

    })
    expect(cmp.html()).toMatchSnapshot()
  })
  it('has correct html structure logged out', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => false
    }
    let store = new Vuex.Store({
      getters
    })
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue
    })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('has correct html structure on jobs index route', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => true
    }
    let store = new Vuex.Store({
      getters
    })
    const $route = {
      name: 'jobs index'
    }
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue,
      mocks: {
        $route
      }
    })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('has correct html structure on jobs new job', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => true
    }
    let store = new Vuex.Store({
      getters
    })
    const $route = {
      name: 'new job'
    }
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue,
      mocks: {
        $route
      }
    })
    expect(cmp.html()).toMatchSnapshot()
  })

  it('logs out', () => {
    // This is a workaround because axios is still the mocked axios
    axios.interceptors = {}
    axios.interceptors.response = {}
    axios.interceptors.response.use = jest.fn()

    let getters = {
      isLoggedIn: () => true
    }
    let actions = {
      logout: jest.fn()
    }
    let store = new Vuex.Store({
      state: {},
      actions,
      getters
    })
    const $route = {
      name: 'jobs index'
    }
    const $router = {
      push: jest.fn()
    }
    let cmp = mount(App, {
      stubs: [
        'router-view',
        'router-link'
      ],
      store,
      localVue,
      mocks: {
        $route,
        $router
      }
    })
    cmp.find('#logout-link').trigger('click')
    expect(actions.logout).toHaveBeenCalled()
    expect($router.push).toHaveBeenCalledWith('/login')
  })
})
