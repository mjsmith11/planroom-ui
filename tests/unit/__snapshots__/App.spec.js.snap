// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has correct html structure logged in 1`] = `
<div id="app">
  <nav class="navbar navbar-expand-lg navbar-dark">
    <router-link-stub to="/jobs" tag="a" class="navbar-brand">
      Planroom
    </router-link-stub> <button type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation" class="navbar-toggler"><span class="navbar-toggler-icon"></span></button>
    <div id="navbarSupportedContent" class="collapse navbar-collapse">
      <ul class="navbar-nav mr-auto">
        <li class="nav-item">
          <router-link-stub to="/jobs" tag="a" class="nav-link">
            Job List
          </router-link-stub>
        </li>
        <li class="nav-item">
          <router-link-stub to="/jobs/new" tag="a" class="nav-link">
            New Job
          </router-link-stub>
        </li>
      </ul>
      <ul class="navbar-nav">
        <li class="nav-item">
          <div id="logout-link" class="nav-link">Logout</div>
        </li>
      </ul>
    </div>
  </nav>
  <router-view-stub></router-view-stub>
</div>
`;

exports[`App has correct html structure logged out 1`] = `
<div id="app">
  <nav class="navbar navbar-expand-lg navbar-dark">
    <router-link-stub to="/login" tag="a" class="navbar-brand">
      Planroom
    </router-link-stub>
    <!---->
    <!---->
  </nav>
  <router-view-stub></router-view-stub>
</div>
`;

exports[`App has correct html structure on jobs index route 1`] = `
<div id="app">
  <nav class="navbar navbar-expand-lg navbar-dark">
    <router-link-stub to="/jobs" tag="a" class="navbar-brand">
      Planroom
    </router-link-stub> <button type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation" class="navbar-toggler"><span class="navbar-toggler-icon"></span></button>
    <div id="navbarSupportedContent" class="collapse navbar-collapse">
      <ul class="navbar-nav mr-auto">
        <li class="nav-item active">
          <router-link-stub to="/jobs" tag="a" class="nav-link">
            Job List
          </router-link-stub>
        </li>
        <li class="nav-item">
          <router-link-stub to="/jobs/new" tag="a" class="nav-link">
            New Job
          </router-link-stub>
        </li>
      </ul>
      <ul class="navbar-nav">
        <li class="nav-item">
          <div id="logout-link" class="nav-link">Logout</div>
        </li>
      </ul>
    </div>
  </nav>
  <router-view-stub></router-view-stub>
</div>
`;

exports[`App has correct html structure on jobs new job 1`] = `
<div id="app">
  <nav class="navbar navbar-expand-lg navbar-dark">
    <router-link-stub to="/jobs" tag="a" class="navbar-brand">
      Planroom
    </router-link-stub> <button type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation" class="navbar-toggler"><span class="navbar-toggler-icon"></span></button>
    <div id="navbarSupportedContent" class="collapse navbar-collapse">
      <ul class="navbar-nav mr-auto">
        <li class="nav-item">
          <router-link-stub to="/jobs" tag="a" class="nav-link">
            Job List
          </router-link-stub>
        </li>
        <li class="nav-item active">
          <router-link-stub to="/jobs/new" tag="a" class="nav-link">
            New Job
          </router-link-stub>
        </li>
      </ul>
      <ul class="navbar-nav">
        <li class="nav-item">
          <div id="logout-link" class="nav-link">Logout</div>
        </li>
      </ul>
    </div>
  </nav>
  <router-view-stub></router-view-stub>
</div>
`;

exports[`App renders for subcontractor 1`] = `
<div id="app">
  <!---->
  <router-view-stub></router-view-stub>
</div>
`;
